Don't tweak boldness of local @font-face fonts already described as bold

fantasai opened the pull request:
#### 21695ca97d9908e30f06c30589efe31c61767b7a
<pre>
Don&apos;t tweak boldness of local @font-face fonts already described as bold
<a href="https://bugs.webkit.org/show_bug.cgi?id=277170">https://bugs.webkit.org/show_bug.cgi?id=277170</a>
<a href="https://rdar.apple.com/130226958">rdar://130226958</a>

Reviewed by Vitor Roriz.

This fixes two problems with lookup of local @font-face fonts described as bold:

1. When using a PostScript name, if the boldness of the font didn&apos;t match the
   descriptor, we would ditch the font and look up a related bold font (which is
   what we do for font-family lookups) instead of using the font as-is.
   See changes in platformFontLookupWithFamily() for this error.

2. When disabling font synthesis because the @font-face descriptors already
   match the request, we fail to plumb that through to local font lookups.
   See changes in CSSFontFaceSource::font() for this error.

* LayoutTests/platform/glib/fast/css/font-face-multiple-faces-expected.txt: Rebaseline.
* LayoutTests/platform/gtk/fast/css/font-face-multiple-faces-expected.png: Rebaseline.
* LayoutTests/fast/css/font-face-multiple-faces.html: Update to match spec expectations.
* LayoutTests/fast/text/postscript-bold-expected.html: More correct references.
* LayoutTests/fast/text/postscript-bold.html: Add test for turning off synthesis.
* LayoutTests/platform/mac/fast/css/font-face-multiple-faces-expected.png: Match new results.
* LayoutTests/platform/mac/fast/css/font-face-multiple-faces-expected.txt: Match new results.
* Source/WebCore/css/CSSFontFaceSource.cpp:
(WebCore::CSSFontFaceSource::load):
(WebCore::CSSFontFaceSource::font):
* Source/WebCore/platform/graphics/FontCache.cpp:
(WebCore::FontCache::cachedFontPlatformData):
(WebCore::FontCache::fontForFamily):
* Source/WebCore/platform/graphics/FontCache.h:
(WebCore::FontCache::fontForFamily):
(WebCore::FontCache::cachedFontPlatformData):
(WebCore::FontCache::createFontPlatformDataForTesting):
* Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
(WebCore::computeNecessarySynthesis):
(WebCore::isAllowlistedFamily):
(WebCore::platformFontLookupWithFamily):
(WebCore::fontDescriptorWithFamilySpecialCase):
(WebCore::fontWithFamily):
(WebCore::FontCache::shouldAutoActivateFontIfNeeded):
(WebCore::autoActivateFont):
(WebCore::lookupFallbackFont):
* Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.h:
(WebCore::computeNecessarySynthesis):
* Source/WebCore/platform/graphics/cocoa/FontFamilySpecificationCoreText.cpp:
(WebCore::FontFamilySpecificationCoreText::fontRanges const):
* Source/WebCore/platform/graphics/freetype/FontCacheFreeType.cpp:
(WebCore::getFontPropertiesFromPattern):
(WebCore::FontCache::createFontPlatformData):
* Source/WebCore/platform/graphics/skia/FontCacheSkia.cpp:
(WebCore::FontCache::createFontPlatformData):
* Source/WebCore/platform/graphics/win/cairo/FontCacheWinCairo.cpp:
(WebCore::FontCache::createFontPlatformData):

Canonical link: <a href="https://commits.webkit.org/281525@main">https://commits.webkit.org/281525@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ee052ad0e7db3534c17cc3eb348f0e64744945c3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/59958 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/39306 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/12510 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/63876 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/10488 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/62087 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/46978 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/10684 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/48585 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7314 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/61988 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/36653 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/51918 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/29426 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/33359 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/9173 "Found 1 new test failure: imported/w3c/web-platform-tests/preload/preload-type-match.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/9407 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/55275 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/9443 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/65608 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/3888 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/9298 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/55934 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/3906 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/51911 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/56084 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/3224 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9020 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/35119 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/36201 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/37289 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/35945 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/b245d9b4b04201502c1254990a1da51444a33206. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/31270

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/9f62e2c11b89d82f70d61531beacc75ac25ff006. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/31270

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/f92666f092f81d61088fd4e26c57660d1d3bbd31. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/31270

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/f8f89d0fad30becd33d2ef0ebe96c4d6b5d6e40e. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/31270

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/ee052ad0e7db3534c17cc3eb348f0e64744945c3. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/31270

fantasai commented:
We don't think the one test failure on WK1 is related: https://ews-build.s3-us-west-2.amazonaws.com/macOS-Ventura-Release-WK1-Tests-EWS/ee052ad0-9173-rerun/imported/w3c/web-platform-tests/preload/preload-type-match-pretty-diff.html

fantasai committed:
Don't tweak boldness of local @font-face fonts already described as bold
https://bugs.webkit.org/show_bug.cgi?id=277170
rdar://130226958

Reviewed by Vitor Roriz.

This fixes two problems with lookup of local @font-face fonts described as bold:

1. When using a PostScript name, if the boldness of the font didn't match the
   descriptor, we would ditch the font and look up a related bold font (which is
   what we do for font-family lookups) instead of using the font as-is.
   See changes in platformFontLookupWithFamily() for this error.

2. When disabling font synthesis because the @font-face descriptors already
   match the request, we fail to plumb that through to local font lookups.
   See changes in CSSFontFaceSource::font() for this error.

* LayoutTests/platform/glib/fast/css/font-face-multiple-faces-expected.txt: Rebaseline.
* LayoutTests/platform/gtk/fast/css/font-face-multiple-faces-expected.png: Rebaseline.
* LayoutTests/fast/css/font-face-multiple-faces.html: Update to match spec expectations.
* LayoutTests/fast/text/postscript-bold-expected.html: More correct references.
* LayoutTests/fast/text/postscript-bold.html: Add test for turning off synthesis.
* LayoutTests/platform/mac/fast/css/font-face-multiple-faces-expected.png: Match new results.
* LayoutTests/platform/mac/fast/css/font-face-multiple-faces-expected.txt: Match new results.
* Source/WebCore/css/CSSFontFaceSource.cpp:
(WebCore::CSSFontFaceSource::load):
(WebCore::CSSFontFaceSource::font):
* Source/WebCore/platform/graphics/FontCache.cpp:
(WebCore::FontCache::cachedFontPlatformData):
(WebCore::FontCache::fontForFamily):
* Source/WebCore/platform/graphics/FontCache.h:
(WebCore::FontCache::fontForFamily):
(WebCore::FontCache::cachedFontPlatformData):
(WebCore::FontCache::createFontPlatformDataForTesting):
* Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
(WebCore::computeNecessarySynthesis):
(WebCore::isAllowlistedFamily):
(WebCore::platformFontLookupWithFamily):
(WebCore::fontDescriptorWithFamilySpecialCase):
(WebCore::fontWithFamily):
(WebCore::FontCache::shouldAutoActivateFontIfNeeded):
(WebCore::autoActivateFont):
(WebCore::lookupFallbackFont):
* Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.h:
(WebCore::computeNecessarySynthesis):
* Source/WebCore/platform/graphics/cocoa/FontFamilySpecificationCoreText.cpp:
(WebCore::FontFamilySpecificationCoreText::fontRanges const):
* Source/WebCore/platform/graphics/freetype/FontCacheFreeType.cpp:
(WebCore::getFontPropertiesFromPattern):
(WebCore::FontCache::createFontPlatformData):
* Source/WebCore/platform/graphics/skia/FontCacheSkia.cpp:
(WebCore::FontCache::createFontPlatformData):
* Source/WebCore/platform/graphics/win/cairo/FontCacheWinCairo.cpp:
(WebCore::FontCache::createFontPlatformData):

Canonical link: https://commits.webkit.org/281525@main

webkit-commit-queue commented:
Committed 281525@main (21695ca97d99): <https://commits.webkit.org/281525@main>

Reviewed commits have been landed. Closing PR #31270 and removing active labels.
